### ISSUE TITLE
Use `macros` with this package to avoid code generation

### ISSUE THREAD

**Issue by JesusHdez960717** [google/reflectable.dart#337]
Is it possible (theoretically at least) to use macros in conjunction with this amazing package to avoid generating code with `build_runner`?
So that the main.reflectable.dart file is generated under the hood.

This idea also implies not having to use `watch` to have the file generated automatically, or wait for the macros to reach a more stable and official point, or the fact that macros also generate code (but in a more autonomous way).

More than an issue, this is a question, but I think this is the most appropriate place to ask it.

Something similar could be possible in the future?

**Comment by eernstg** [comment]
Interesting question, and thanks for the kind words about the package!

I've had similar things in mind for a long time, but it isn't a perfect match.

First, reflectable is based on a whole-program analysis, and it includes some quantification mechanisms where the ability to inspect the entire program is crucial. For instance, it's possible to specify a reflector (using a subclass of the class `Reflectable`) which is applied to all non-bottom subtypes of each class which is explicitly included. So if `reflector` is a constant denoting that reflector then we would declare `@reflector class MyClass ...` and this would cause `MyClass` as well as all subtypes of `MyClass` in the program to be included in the kind of reflection support that said reflector specifies. Other quantifications allow for specifying that a given reflector must cover a specified class in some other library (which makes it possible to cover classes like `int`, or in general covering classes that we can't edit).

A macro can modify a declaration if it occurs as metadata on that declaration itself. I don't think the macro system will allow for quantification over the entire program, or modifying declarations in other libraries.

When macros have been finalized and released we will know more about how difficult it would be to do something like a re-implementation of reflectable using macros. Right now I think the answer is "maybe". 😄

**Comment by JesusHdez960717** [comment]
You are absolutely right, we are working on a completely theoretical basis, we will have to wait for the macros to reach a stable point so that we can assess whether the possibility exists.

Even so, I see that at least it has been thought about, and although it is not a perfect match, I think that something can be done so that they work together and strengthen each other (macros and reflectable).

For now I am happy knowing that the answer is: 'maybe'

My doubt is satisfied, thank you very much

I will then close the issue
